cli/tests/test_silo_idp_saml.rs

```diff
@@ -78,3 +82,130 @@ fn test_silo_idp_saml_create() {
+            .idp_entity_id("17")
+            .idp_metadata_source(IdpMetadataSource::Base64EncodedXml { data: encoded_xml })
+            .technical_contact_email("anyone@but.adam")
+            .signing_keypair(DerEncodedKeyPair {
+                private_key: encoded_key,
+                public_cert: encoded_cert,
+            })
+            .try_into()
+            .unwrap();
+        when.body(&body);
+        then.created(&SamlIdentityProvider {
+            acs_url: body.acs_url.clone(),
+            description: body.description.clone(),
+            idp_entity_id: body.idp_entity_id.clone(),
+            name: body.name.clone(),
+            slo_url: body.slo_url.clone(),
+            sp_client_id: body.sp_client_id.clone(),
+            technical_contact_email: body.technical_contact_email,
+            public_cert: body.signing_keypair.map(|kp| kp.public_cert),
+            ..JsonMock::mock_value(&mut src).unwrap()
+        });
+    });
+
+    Command::cargo_bin("oxide")
+        .unwrap()
+        .env("RUST_BACKTRACE", "1")
+        .env("OXIDE_HOST", server.url(""))
+        .env("OXIDE_TOKEN", "fake-token")
+        .arg("silo")
+        .arg("idp")
+        .arg("saml")
+        .arg("create")
+        .arg("--name")
+        .arg("samlrific")
+        .arg("--description")
+        .arg("for all the security I guess")
+        .arg("--silo")
+        .arg("mysilo")
+        .arg("--acs-url")
+        .arg("http://nope.nope")
+        .arg("--slo-url")
+        .arg("http://nope.nope")
+        .arg("--sp-client-id")
+        .arg("17")
+        .arg("--idp-entity-id")
+        .arg("17")
+        .arg("--metadata-value")
+        .arg(&xml_path)
+        .arg("--technical-contact-email")
+        .arg("anyone@but.adam")
+        .arg("--private-key")
+        .arg(&key_path)
+        .arg("--public-cert")
+        .arg(&cert_path)
+        .assert()
+        .success()
+        .stdout(expectorate::eq_file_or_panic(
+            "tests/data/test_silo_idp_saml_create_with_files.stdout",
+        ));
+
+    mock.assert();
+}
+
+#[test]
+fn test_silo_idp_saml_create_requires_key_and_cert() {
+    Command::cargo_bin("oxide")
+        .unwrap()
+        .env("RUST_BACKTRACE", "1")
+        .env("OXIDE_HOST", "invalid.url")
+        .env("OXIDE_TOKEN", "fake-token")
+        .arg("silo")
+        .arg("idp")
+        .arg("saml")
+        .arg("create")
+        .arg("--name")
+        .arg("samlrific")
+        .arg("--description")
+        .arg("for all the security I guess")
+        .arg("--silo")
+        .arg("mysilo")
+        .arg("--acs-url")
+        .arg("http://nope.nope")
+        .arg("--slo-url")
+        .arg("http://nope.nope")
+        .arg("--sp-client-id")
+        .arg("17")
+        .arg("--idp-entity-id")
+        .arg("17")
+        .arg("--metadata-url")
+        .arg("http://nope.nope")
+        .arg("--technical-contact-email")
+        .arg("anyone@but.adam")
+        .arg("--private-key")
```

Comment:
this fails because this guy didn't have his friend?

Reply:
Yup, that's correct. Do you want a comment here to be explicit?

Reply:
I'm thinking of myself in the future, breaking this, but not noticing.